include/boost/asio/windows_sspi/stream.hpp

```diff
@@ -119,14 +162,67 @@ template <typename NextLayer, typename ConstBufferSequence> struct async_write_i
                                              net::buffer(m_message),
                                              net::transfer_exactly(m_message.size()),
                                              std::move(self));
-      self.complete(ec, net::buffer_size(m_buffer));
+      self.complete(ec, m_size_encrypted);
     }
   }
 
 private:
   NextLayer& m_next_layer;
   ConstBufferSequence m_buffer;
-  std::shared_ptr<sspi_impl> m_sspi;
+  std::shared_ptr<sspi_impl> m_sspi_impl;
+  std::vector<char> m_message;
+  size_t m_size_encrypted{0};
+};
+
+// TODO: Move away from this file
+template <typename NextLayer, typename MutableBufferSequence> struct async_read_impl : boost::asio::coroutine {
+  async_read_impl(NextLayer& next_layer, const MutableBufferSequence& buffer, std::shared_ptr<sspi_impl> sspi_impl)
+    : m_next_layer(next_layer)
+    , m_buffer(buffer)
+    , m_sspi_impl(std::move(sspi_impl)) {
+  }
+
+  template <typename Self> void operator()(Self& self, boost::system::error_code ec = {}, std::size_t length = 0) {
+    BOOST_ASIO_CORO_REENTER(*this) {
+      while(m_sspi_impl->decrypted_data.empty()) {
+        // TODO: Find some way to make the sspi_impl, the decrypt
+        // function or something else be responsible for keeping track
+        // of state and buffer(s)
+        m_message.resize(0x10000 - m_sspi_impl->encrypted_data.size());
```

Comment:
Consider an assert to make sure the calculation does not underflow

Reply:
Good point. That's actually a real risk in this case. Of course, I'd rather handle all of this differently, but that doesn't make the overflow less of an issue currently.